clvm-traits/docs/derive_macros.md

```diff
@@ -0,0 +1,218 @@
+use clvmr::Allocator;
+use clvm_traits::{ToClvm, FromClvm};
+
+#[derive(Debug, PartialEq, Eq, ToClvm, FromClvm)]
+#[clvm(list)]
+struct PasswordSolution {
+    password: String,
+}
+
+let solution = PasswordSolution {
+    password: "Hello".into(),
+};
+
+let a = &mut Allocator::new();
+let ptr = solution.to_clvm(a).unwrap();
+assert_eq!(PasswordSolution::from_clvm(a, ptr).unwrap(), solution);
+```
+
+### Curry
+
+This represents the argument part of a curried CLVM program. Currying is a method of partially
+applying some of the arguments without immediately calling the function.
+
+For example, `(A, B, C)` is encoded as `(c (q . A) (c (q . B) (c (q . C) 1)))`. Note that the
+arguments are quoted and terminated with `1`, which is how partial application is implemented in CLVM.
+
+You can read more about currying on the [Chia blockchain documentation](https://docs.chia.net/guides/chialisp-currying).
+
+Note that the following code is for example purposes only and is not indicative of how to create a secure program.
+Using a password like shown in this example is an insecure method of locking coins, but it's effective for learning.
+
+```rust
+use clvmr::Allocator;
+use clvm_traits::{ToClvm, FromClvm};
+
+#[derive(Debug, PartialEq, Eq, ToClvm, FromClvm)]
+#[clvm(curry)]
+struct PasswordArgs {
+    password: String,
+}
+
+let args = PasswordArgs {
+    password: "Hello".into(),
+};
+
+let a = &mut Allocator::new();
+let ptr = args.to_clvm(a).unwrap();
+assert_eq!(PasswordArgs::from_clvm(a, ptr).unwrap(), args);
+```
+
+## Enums
+
+In Rust, enums contain a discriminant, a value used to distinguish between each variant of the enum.
+In most cases, the CLVM representation of the enum will need to contain this discriminant as the first argument.
+For convenience, this is the behavior when deriving `ToClvm` and `FromClvm` for enums by default.
+
+### Simple Example
+
+In this example, since the `tuple` representation is used and the only values are the discriminants, the variants will be encoded as an atom.
+Discriminants default to the `isize` type and the first value is `0`. Subsequent values are incremented by `1` by default.
+
+```rust
+use clvmr::Allocator;
+use clvm_traits::{ToClvm, FromClvm};
+
+#[derive(Debug, PartialEq, Eq, ToClvm, FromClvm)]
+#[clvm(tuple)]
+enum Status {
+    Pending,
+    Completed,
+}
+
+let status = Status::Pending;
+
+let a = &mut Allocator::new();
+let ptr = status.to_clvm(a).unwrap();
+assert_eq!(Status::from_clvm(a, ptr).unwrap(), status);
+```
+
+### Custom Discriminator
+
+It's possible to override both the type of the discriminator, and the value.
+The `#[repr(...)]` attribute is used by the Rust compiler to allow overriding the discriminator type.
+As such, this attribute is also used to change the underlying type used to serialize and deserialize discriminator values.
+
+```rust
+use clvmr::Allocator;
+use clvm_traits::{ToClvm, FromClvm};
+
+#[derive(Debug, PartialEq, Eq, ToClvm, FromClvm)]
+#[clvm(tuple)]
```

Comment:
Yes, you must specify one of the 3 representations when deriving the macros. There isn't a default, since it wouldn't necessarily be clear which one would be the default.